File: day6.py
""" Solution for Day 7 """

import os

class Body:
    """ Represents an orbiting body.

    Args:
        name: the name of the body
    """

    def __init__(self, name: str):
        self.satellites = []
        self.name = name

    def add_satellite(self, body: "Body"):
        """ Add a satellite to this body """
        self.satellites.append(body)

    def count_orbits(self, depth: int) -> int:
        """ Count the number of direct and indirect orbits """
        if not self.satellites:
            return depth

        orbits = depth + sum([satellite.count_orbits(depth + 1)
                              for satellite in self.satellites])
        return orbits

    def find(self, name: str) -> int:
        """ Find a named body in this body's orbit list """
        if self.name == name:
            return 1

        distances = list(filter(None, [body.find(name) for body in self.satellites]))
        if distances:
            return min(distances) + 1

        return None

    def find_shortest_transfer(self, name0: str, name1: str) -> int:
        """ Find the shortest orbital transfer between two bodies """
        if not self.satellites:
            return None

        distances0 = list(filter(None, [body.find(name0)
                                        for body in self.satellites]))
        distances1 = list(filter(None, [body.find(name1)
                                        for body in self.satellites]))

        if not(distances0 and distances1):
            return None

        path_distance = list(filter(None, [body.find_shortest_transfer(name0, name1)
                                           for body in self.satellites]))
        if path_distance:
            return min(path_distance)

        return min(distances0) + min(distances1) - 2

    @staticmethod
    def parse(orbits: str) -> "Body":
        """ Parse a body and its orbits from a string """
        com = Body("COM")
        lookup = {com.name: com}

        for orbit in orbits.split("\n"):
            body0, body1 = orbit.split(')')
            if body0 not in lookup:
                lookup[body0] = Body(body0)

            if body1 not in lookup:
                lookup[body1] = Body(body1)

            lookup[body0].add_satellite(lookup[body1])

        return com


TEST0 = """COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L"""

TEST1 = """COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L
K)YOU
I)SAN"""


def test_count_orbits():
    """ Test the count orbits method """
    com = Body.parse(TEST0)
    assert com.count_orbits(0) == 42


def test_shortest_transfer():
    """ Test the shortest transfer method """
    com = Body.parse(TEST1)
    assert com.find_shortest_transfer("YOU", "SAN") == 4


def _main():
    with open(os.path.join("inputs", "day6.txt")) as file:
        com = Body.parse(file.read())

    print("part1:", com.count_orbits(0))
    print("part2:", com.find_shortest_transfer("YOU", "SAN"))


if __name__ == "__main__":
    _main()

File: haskell/day6.hs
import Data.Maybe
import qualified Data.Map as Map
import Test.HUnit
import Debug.Trace
import System.IO
import System.FilePath

data Body = Body { 
    name :: String,
    satellites :: [Body]
} deriving Show

-- TODO use map, as it doesn't seem possible to build a tree?

mapJust :: (Body -> Maybe a) -> [Body] -> [Maybe a]
mapJust func bodies = filter isJust (map func bodies)

countOrbits :: Int -> Body -> Int
countOrbits depth (Body name satellites)
    | null satellites = depth
    | otherwise = depth + sum (map (countOrbits (depth + 1)) satellites)
    
distanceTo :: String -> Body -> Maybe Int
distanceTo target (Body name satellites)
    | name == target = Just 0
    | null satellites = Nothing
    | otherwise = do
        let distances = mapJust (distanceTo target) satellites
        if null distances then
            Nothing
        else
            Just(+1) <*> minimum distances

minJust :: [Maybe Int] -> Int
minJust values = fromJust (minimum values)

-- TODO this is horrible. A much better solution would do something like:
-- 1. find the path down the tree to each thing, record distance to node
-- 2. for all elements in both lists, find the min sum

findShortestTransfer :: String -> String -> Body -> Maybe Int
findShortestTransfer name0 name1 (Body name satellites)
    | null satellites = Nothing
    | otherwise = do
        let path_distance = filter (/=Nothing) (map (findShortestTransfer name0 name1) satellites)

        if not (null path_distance) then
            minimum path_distance
        else do
            let distances0 = mapJust (distanceTo name0) satellites
            let distances1 = mapJust (distanceTo name1) satellites

            if null distances0 || null distances1 then
                Nothing
            else 
                Just (minJust distances0 + minJust distances1)

prep :: Char -> Char
prep char
    | char == ')' = ' '
    | otherwise = char

splitOrbit :: String -> Map.Map String [String]
splitOrbit orbit = do
    let name:satellites = words (map prep orbit)
    Map.fromList [(name, satellites)]

parseOrbits :: [String] -> Map.Map String [String]
parseOrbits orbits = foldl (Map.unionWith (++)) Map.empty (map splitOrbit orbits)

buildTree :: Map.Map String [String] -> String -> Body
buildTree orbitMap name 
    | Map.member name orbitMap = Body name (map (buildTree orbitMap) (fromJust (Map.lookup name orbitMap)))
    | otherwise = Body name []

parseCOM :: [String] -> Body
parseCOM orbits = buildTree (parseOrbits orbits) "COM"

testOrbits0 = [
    "COM)B",
    "B)C",
    "C)D",
    "D)E",
    "E)F",
    "B)G",
    "G)H",
    "D)I",
    "E)J",
    "J)K",
    "K)L"]

testOrbits1 = [
    "COM)B",
    "B)C",
    "C)D",
    "D)E",
    "E)F",
    "B)G",
    "G)H",
    "D)I",
    "E)J",
    "J)K",
    "K)L",
    "K)YOU",
    "I)SAN"]

tests = test [
    "testCountOrbits" ~: 42 ~=? countOrbits 0 (parseCOM testOrbits0),
    "testFindShortestTransfer" ~: 4 ~=? fromJust (findShortestTransfer "YOU" "SAN" (parseCOM testOrbits1))]

main = do
    let path = ".." ++ [pathSeparator] ++ "inputs" ++ [pathSeparator] ++ "day6.txt"
    contents <- readFile path
    let orbits = lines contents
        com = parseCOM orbits

    putStrLn ("Part 1: " ++ show (countOrbits 0 com))

    putStrLn ("Part 2: " ++ show (fromJust (findShortestTransfer "YOU" "SAN" com)))
